src/protocol/common/hex.rs

```diff
@@ -168,7 +168,42 @@ pub fn decode_hex_buf(base_buf: &mut [u8]) -> Result<&mut [u8], DecodeHexBufErro
     Ok(&mut base_buf[..decoded_len + odd_adust])
 }
 
-#[allow(dead_code)]
+#[derive(Debug)]
+pub enum DecodeBinBufError {
+    UnexpectedEnd,
+}
+
+/// Decode GDB escaped binary bytes into origin bytes _in place_.
+pub fn decode_bin_buf(buf: &mut [u8]) -> Result<&mut [u8], DecodeBinBufError> {
+    use DecodeBinBufError::*;
+    let mut i = 0;
+    let mut j = 0;
+    let len = buf.len();
+    while i < len && j < len {
+        if buf[i] == b'}' {
+            if i + 1 >= len {
+                return Err(UnexpectedEnd);
+            } else {
+                buf[j] = buf[i + 1] ^ 0x20;
+                i += 1;
+            }
+        } else {
+            buf[j] = buf[i];
+        }
+        i += 1;
+        j += 1;
+    }
+
+    // SAFETY: by inspection, the value of j will never exceed buf.len().
+    // Unfortunately, the LLVM optimizer isn't smart enough to see this, so
+    // we have to manually elide the bounds check...
+    if cfg!(feature = "paranoid_unsafe") {
+        Ok(&mut buf[..j])
+    } else {
+        unsafe { Ok(buf.get_unchecked_mut(..j)) }
```

Comment:
ahh, my bad: `debug_assert!(false)` is obviously incorrect, but we should add a `debug_assert!(j <= buf.len())` before calling `get_unchecked_mut` here.

Reply:
Why there are no `debug_assert` in `decode_hex_buf`?

Reply:
no reason aside from the fact that I forgot to add them. I should probably insert them at some point, just in case...